src/main/java/org/candlepin/subscriptions/marketplace/MarketplaceProducer.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright (c) 2021 Red Hat, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ * Red Hat trademarks are not licensed under GPLv3. No permission is
+ * granted to use or replicate Red Hat trademarks that are incorporated
+ * in this software or its documentation.
+ */
+package org.candlepin.subscriptions.marketplace;
+
+import org.candlepin.subscriptions.exception.ErrorCode;
+import org.candlepin.subscriptions.exception.SubscriptionsException;
+import org.candlepin.subscriptions.marketplace.api.model.StatusResponse;
+import org.candlepin.subscriptions.marketplace.api.model.UsageEvent;
+import org.candlepin.subscriptions.marketplace.api.model.UsageRequest;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Qualifier;
+import org.springframework.retry.support.RetryTemplate;
+import org.springframework.stereotype.Service;
+
+import io.micrometer.core.annotation.Timed;
+
+import java.util.stream.Collectors;
+
+import javax.ws.rs.core.Response;
+
+/**
+ * Component that is responsible for emitting usage info to Marketplace, including handling retries.
+ */
+@Service
+public class MarketplaceProducer {
+
+    private static final Logger log = LoggerFactory.getLogger(MarketplaceProducer.class);
+
+    private final MarketplaceService marketplaceService;
+    private final RetryTemplate retryTemplate;
+
+    @Autowired
+    MarketplaceProducer(MarketplaceService marketplaceService,
+        @Qualifier("marketplaceRetryTemplate") RetryTemplate retryTemplate) {
+        this.marketplaceService = marketplaceService;
+        this.retryTemplate = retryTemplate;
+    }
+
+    @Timed("rhsm-subscriptions.marketplace.usage")
+    public StatusResponse submitUsageRequest(UsageRequest usageRequest) {
+        // NOTE: https://issues.redhat.com/browse/ENT-3609 will address failures
+        return retryTemplate.execute(context -> tryRequest(usageRequest));
+    }
+
+    @Timed("rhsm-subscriptions.marketplace.usage.request")
+    private StatusResponse tryRequest(UsageRequest usageRequest) {
```

Comment:
Yeah, when I ran through the test steps, here's what I got from prometheus:

```% curl -s http://localhost:8080/actuator/prometheus | grep marketplace
# HELP rhsm_subscriptions_marketplace_usage_seconds  
# TYPE rhsm_subscriptions_marketplace_usage_seconds summary
rhsm_subscriptions_marketplace_usage_seconds_count{class="org.candlepin.subscriptions.marketplace.MarketplaceProducer",exception="none",method="submitUsageRequest",} 1.0
rhsm_subscriptions_marketplace_usage_seconds_sum{class="org.candlepin.subscriptions.marketplace.MarketplaceProducer",exception="none",method="submitUsageRequest",} 0.975712562
# HELP rhsm_subscriptions_marketplace_usage_seconds_max  
# TYPE rhsm_subscriptions_marketplace_usage_seconds_max gauge
rhsm_subscriptions_marketplace_usage_seconds_max{class="org.candlepin.subscriptions.marketplace.MarketplaceProducer",exception="none",method="submitUsageRequest",} 0.975712562
```

So I don't think that `tryRequest` is getting timed since there's nothing there for `usage.request`.

Reply:
So, I could have sworn I had both metrics working... maybe before some refactoring? Anyways, I moved the annotation to `MarketplaceService` instead, since that actually makes the request. Thanks for the thoroughness.

FYI, I also renamed `rhsm-subscriptions.marketplace.usage` to `rhsm-subscriptions.marketplace.usage.submission` because that felt better than having `usage` and `usage.request`...

Reply:
:moneybag: :+1:

Reply:
I can confirm that both metrics are there now:

```
% curl -s http://localhost:8080/actuator/prometheus | grep marketplace
# HELP rhsm_subscriptions_marketplace_usage_submission_seconds  
# TYPE rhsm_subscriptions_marketplace_usage_submission_seconds summary
rhsm_subscriptions_marketplace_usage_submission_seconds_count{class="org.candlepin.subscriptions.marketplace.MarketplaceProducer",exception="none",method="submitUsageRequest",} 1.0
rhsm_subscriptions_marketplace_usage_submission_seconds_sum{class="org.candlepin.subscriptions.marketplace.MarketplaceProducer",exception="none",method="submitUsageRequest",} 1.193515685
# HELP rhsm_subscriptions_marketplace_usage_submission_seconds_max  
# TYPE rhsm_subscriptions_marketplace_usage_submission_seconds_max gauge
rhsm_subscriptions_marketplace_usage_submission_seconds_max{class="org.candlepin.subscriptions.marketplace.MarketplaceProducer",exception="none",method="submitUsageRequest",} 1.193515685
# HELP rhsm_subscriptions_marketplace_usage_request_seconds  
# TYPE rhsm_subscriptions_marketplace_usage_request_seconds summary
rhsm_subscriptions_marketplace_usage_request_seconds_count{class="org.candlepin.subscriptions.marketplace.MarketplaceService",exception="none",method="submitUsageEvents",} 1.0
rhsm_subscriptions_marketplace_usage_request_seconds_sum{class="org.candlepin.subscriptions.marketplace.MarketplaceService",exception="none",method="submitUsageEvents",} 1.183517352
# HELP rhsm_subscriptions_marketplace_usage_request_seconds_max  
# TYPE rhsm_subscriptions_marketplace_usage_request_seconds_max gauge
rhsm_subscriptions_marketplace_usage_request_seconds_max{class="org.candlepin.subscriptions.marketplace.MarketplaceService",exception="none",method="submitUsageEvents",} 1.183517352
```